### ISSUE TITLE
the velocity field integration method 'quadrature'

### ISSUE THREAD

**Issue by shuo-shuo-zhang** [voxelmorph/voxelmorph#79]
Dear Dr. Dalca,
I'm sorry to bother you again. Recently, I am working on the stationary velocity field integration method, I notice that you mentioned 3 methods, including scaling and squaring, quadrature and ode. Is the quadrature method the forward Euler method? Because it seems like they have similar integration process, thank you very much.

And you mentioned the inverse-consistency error of the forward and the inverse deformations, I want to know how to compute it, does your code give that part? Is the composition of the two deformations the same with that of the scaling and squaring method? If they are the same composition, then I know where the find the code, it is the 'transform' part of the integrate_vec. 

Thank you very much!

**Comment by adalca** [comment]
I suppose it is, but you'll get more intuition of how it works by looking here; https://en.wikipedia.org/wiki/Numerical_integration#Quadrature_rules_based_on_interpolating_functions

regarding inverse consistency, I'm not sure where exactly you are referring. There are several ways that people compute this -- one is to compose the two deformation fields (forward and inverse) and see how close you are to identity (0 displacement). Yes, the `transform` part of the code will help with this. Essentially, if you have a `fwd` and a `bwd` transform, you want to do something like warp the `fwd` model with the `bwd` model, and then add the `bwd` model to this -- the result should be close to 0 if the models are true inverses.

**Comment by shuo-shuo-zhang** [comment]
Re: Thank you very much, I understand what you mean. Just have one more thing to confirm.

You mentioned in your paper that the effects of integration steps on the registration accuracy, the registration time, the number of folding voxels and the inverse consistency error.   You said "during training, we hold the scaling and squaring steps fixed. However, this number can be varied at test time." Does this mean that we can train the model with a fixed steps like nb_steps = 7, and test this trained model with different steps like nb_steps = 5,6,7,8,9 ? I mean, should the parameter nb_steps in the training model be  necessarily the same with that of the test model? Because if we use one parameter value in the training process while use a different value for the same parameter in the test process, I think that will make the test results inaccurate. So maybe I should train 5 models with 5 values of nb_steps and test them respectively. Does this make any sense? I did train one model with nb_steps = 8 and tested this model with 5 different values, 5,6,7,8,9, and found that the dice scores are almost the same, and as the parameter increases  from 5 to 7, the number of  folding voxels increases a little instead of decreasing like your result and the registration time decreases a little, while the  inverse consistency error follows your result. So I think maybe I should use different values of the parameter nb_steps to train models. 

Think you very much and look forward to your reply !

**Comment by adalca** [comment]
yes, you can train with `nb_steps=7` (something sufficiently larger), and test with a different `nb_steps` value (usually you would want this the same or larger at test time, otherwise i don't see a point of changing it) -- that's actually what we show in the paper -- the results don't really become less accurate, but you should get better Jacobian determinant values! This is because the number of steps will affect the accuracy of the integration. 

I don't think you should train 5 models. In your model trained with 8 steps, can you plot the Jacobian determinant values for `nb_steps = 6,7,8,9` over, say, 50 subjects? If the number of folding voxels go *up*, that's an issue, we never observed something like this.

**Comment by shuo-shuo-zhang** [comment]
I'm so sorry for the late response, I use 37 subjects to get the mean of folding voxels' number  when fix the number of integration steps and plot the variation trend of the folding voxels' number as the number of steps increases. The figure is shown as follows.
![folding_voxels](https://user-images.githubusercontent.com/50296395/62097786-a2699380-b2ba-11e9-89ae-07758bb22d8d.png)
 Thank you very much!

**Comment by shuo-shuo-zhang** [comment]
I'm sorry, forgot to ask, you suggest me  train with a smaller  **nb_steps**， such as 5  and test with some larger **nb_steps**， such as 5,6,7,8,9  rather than train with 8 and test with 5,6 and 7. Is this the rule, maybe in this way, I can get a down trend as the nb_steps goes up.

Thank you very much!

**Comment by adalca** [comment]
Whoah, that is a lot of folding voxels -- can you show the standard deviations on the graph? also, what images are these, how did you train, etc?

I think you should train with a high number of steps if you can, I was just saying that in some cases, users might *need* to use a low number of steps because of memory considerations at training.

**Comment by Barnonewdm** [comment]
Thanks for @adalca's response. My experience is consistent with @shuo-shuo-zhang's. In my understanding, the exact integration method "scaling and squaring" is doing scaling and composition. Most importantly, this step is not an identity.  The displacement norm will decrease with the increase of integration number (nb_steps in code). If the nb_steps is large enough, i.e. 150, the displacement after integration is almost zero (the deformation field should be identity). Therefore, I do agree that the folding points should decrease with the number of integration. But the figure showed by @shuo-shuo-zhang reveals one more important issue that folding performance is not sensitive to the number of integration.

**Comment by adalca** [comment]
Thanks for the input @Barnonewdm, @shuo-shuo-zhang .

Can you both confirm that you are using `use_miccai_int=False` in `miccai2018_net`, that is, the default?

@Barnonewdm, your interpretation of scaling and squaring is not quite correct --> you can think of *scaling and squaring* as approximating the integral \phi = int(v), and the number of steps will simply affect how close this approximation is. Therefore, with very high  `nb_steps`, the resulting displacement will be close to what you would get from the smooth integral -- so not close to 0, but close to the analytical integration solution, and should have fewer folding voxels (0 in the limit of infinite steps). We do indeed find this empirically in the paper (see Fig 10 in https://arxiv.org/pdf/1903.03545.pdf). 

I'm not quite sure why the behaviour in the figure of @shuo-shuo-zhang is as it is, and I'm happy to help debug it. The first thing to make sure is that the `use_miccai_int` variable is set to `False`. Next, it would be great for @shuo-shuo-zhang to explain how he built that graph, exactly.

**Comment by shuo-shuo-zhang** [comment]
I have to prepare my answer and images, so I need some time, when I get them prepared, I'll put the details here.

Thank you very much!

**Comment by shuo-shuo-zhang** [comment]
Dear Dr. Dalca,
I'm sure I set the  use_miccai_int variable to False when training. I use the Colin27 template to align the ADNI images and perform almost the same  pre-processing steps with you except the histogram matching I use to reduce the gray scale gap of the Colin template and ADNI images.  For training details, I use the mse loss for warped image and fixed image, and L2 regularization loss for velocity field to make it smooth, the hyper-parameter ratio for the two is about 17. I use the Adam optimizer and train about 800 epochs and with ss of 8 nb_steps.  But during training, the standard deviation for each integration step is  very large,  the number I have shown in the figure above is a mean of 37 subjects and is almost 380, while the max is about 1000 and the min is 0, so the standard deviation is very large.

That is all that comes to mind and tell me if I missed something. Thank you very much!

**Comment by adalca** [comment]
Can you specify what you mean by L2 loss on the deformation field? This sounds like you changed the loss function from the default miccai loss?

**Comment by shuo-shuo-zhang** [comment]
Dear Dr. Dalca,
I just tried to increase the velocity field regularization parameter a lot, and the number of folding voxels decreased to almost 0, however the dice score also went down a lot. I don't know which one is more important, the dice score or the number of folding voxels, which one should I make sure is more accurate, the dice or the folding voxels? 

Thank you very much!

**Comment by shuo-shuo-zhang** [comment]
yes, I changed the kl_loss to the L2 regularization loss which a lot of people use to make the deformation field smooth. I think it is similar to the kl_loss.

**Comment by adalca** [comment]
okay, you might want to mention the changes you made up front when you ask a question, it helps us figure out what's going on :)

You should be able to achieve both high Dice and nearly no folding

Let's see -- my question was more what do you apply L2 to? the velocity mean? the velocity sample? The problem is that without a KL loss, I'm not quite sure why the variance wouldn't go to ~0 -- the gradients would point in that direction. Thus, I think it's best if you explain in detail what you changed.

Also, if you're changing losses and data, you might as well change MSE to NCC -- this will be more robust to changes in the data you have (but you'll need to tune parameters)
